src/datasources/asset/AssetDataSource.ts

```diff
@@ -0,0 +1,111 @@
+import {
+  DataFrameDTO,
+  DataQueryRequest,
+  DataSourceInstanceSettings,
+  TestDataSourceResponse,
+} from '@grafana/data';
+import { BackendSrv, getBackendSrv, getTemplateSrv, TemplateSrv } from '@grafana/runtime';
+import { DataSourceBase } from 'core/DataSourceBase';
+import {
+  AssetFilterProperties,
+  AssetModel, AssetQuery,
+  AssetQueryType,
+  AssetsResponse,
+} from './types';
+import { getWorkspaceName, replaceVariables } from "../../core/utils";
+import { SystemMetadata } from "../system/types";
+import { defaultOrderBy, defaultProjection } from "../system/constants";
+
+export class AssetDataSource extends DataSourceBase<AssetQuery> {
+  constructor(
+    readonly instanceSettings: DataSourceInstanceSettings,
+    readonly backendSrv: BackendSrv = getBackendSrv(),
+    readonly templateSrv: TemplateSrv = getTemplateSrv()
+  ) {
+    super(instanceSettings, backendSrv, templateSrv);
+  }
+
+  baseUrl = this.instanceSettings.url + '/niapm/v1';
+
+  defaultQuery = {
+    type: AssetQueryType.Metadata,
+    workspace: '',
+    minionIds: [],
+  };
+
+  async runQuery(query: AssetQuery, options: DataQueryRequest): Promise<DataFrameDTO> {
+    return await this.processMetadataQuery(query)
+  }
+
+  async processMetadataQuery(query: AssetQuery) {
+    const result: DataFrameDTO = { refId: query.refId, fields: [] };
+    let workspaceId = this.templateSrv.replace(query.workspace);
+    const minionIds = replaceVariables(query.minionIds, this.templateSrv);
+    const conditions = [];
+    if (workspaceId) {
+      conditions.push(`workspace = "${workspaceId}"`);
+    }
+    if (minionIds.length) {
+      const systemsCondition = minionIds.map(id => `${AssetFilterProperties.LocationMinionId} = "${id}"`)
+      conditions.push(`(${systemsCondition.join(' or ')})`);
+    }
+    const assetFilter = conditions.join(' and ');
+    const assets: AssetModel[] = await this.queryAssets(assetFilter);
+    const workspaces = await this.getWorkspaces();
+    result.fields = [
+      { name: 'id', values: assets.map(a => a.id) },
+      { name: 'name', values: assets.map(a => a.name) },
+      { name: 'model name', values: assets.map(a => a.modelName) },
+      { name: 'serial number', values: assets.map(a => a.serialNumber) },
+      { name: 'bus type', values: assets.map(a => a.busType) },
+      { name: 'asset type', values: assets.map(a => a.assetType) },
+      { name: 'is NI asset', values: assets.map(a => a.isNIAsset) },
+      { name: 'calibration status', values: assets.map(a => a.calibrationStatus) },
+      { name: 'is system controller', values: assets.map(a => a.isSystemController) },
+      { name: 'last updated timestamp', values: assets.map(a => a.lastUpdatedTimestamp) },
+      { name: 'minionId', values: assets.map(a => a.location.minionId) },
+      { name: 'parent name', values: assets.map(a => a.location.parent) },
+      { name: 'system name', values: assets.map(a => a.location.systemName) },
+      {
```

Comment:
Is this something necessary now? It only uses NextRecommendedDate and does not take into account the NextCustomDueDate (overwrite for the first one). We plan to add better support for this

Reply:
ok, I'll delete this line